### PR TITLE
Better reporting for query cache hits and misses for OTEL.

### LINKED ISSUE #33959: Better reporting for compiled_query_cache_hit_rate metric
https://github.com/dotnet/docs/pull/41358#discussion_r1634373713

### PR DESCRIPTION
**Pull request by cincuranet**
Fixes #33959

### CODE REVIEW

File: src/EFCore/Infrastructure/EntityFrameworkMetricsData.cs [hunk 1]
```diff
@@ -138,7 +150,7 @@ private struct CacheInfo
                 : new CacheInfo { _all = Interlocked.Read(ref _all) };
             return CalculateHitsMissesHitRateImpl(clone);
 
-            static (int hits, int misses, double hitRate) CalculateHitsMissesHitRateImpl(CacheInfo cacheInfo)
+            static (int hits, int misses, double hitRate) CalculateHitsMissesHitRateImpl(in CacheInfo cacheInfo)
```
Comment (roji) [inline src/EFCore/Infrastructure/EntityFrameworkMetricsData.cs:153]: What's this about? :)

Quoting from [this article](https://devblogs.microsoft.com/premier-developer/the-in-modifier-and-the-readonly-structs-in-c/) (which is recommended reading):

> you should never pass a non-readonly struct as in parameter

**Review by roji — APPROVED** [review]
LGTM except for those that little `in` that got added there

File: src/EFCore/Infrastructure/EntityFrameworkMetricsData.cs (hunk 1 above)
Comment (cincuranet) [reply src/EFCore/Infrastructure/EntityFrameworkMetricsData.cs]: Leftover from playing. Removed.

File: src/EFCore/Infrastructure/EntityFrameworkEventSource.cs [hunk 2]
```diff
@@ -168,7 +168,7 @@ protected override void OnEventCommand(EventCommandEventArgs command)
             _compiledQueryCacheHitRateCounter ??= new PollingCounter(
                 "compiled-query-cache-hit-rate",
                 this,
-                () => EntityFrameworkMetricsData.GetCompiledQueryCacheHitsMissesHitRate().hitRate)
+                () => EntityFrameworkMetricsData.GetCompiledQueryCacheHitsMissesHitRateEventSource().hitRate)
```
[the remark this answers is not in the document]
Comment (cincuranet) [reply src/EFCore/Infrastructure/EntityFrameworkEventSource.cs:171]: Can you elaborate?

[the remark this answers is not in the document]
Comment (cincuranet) [reply src/EFCore/Infrastructure/EntityFrameworkEventSource.cs]: Be my guest.

**Pull request merged**